File: nova/mobile/hybrid/device_discovery.py
"""Device Discovery — Nova Core TCP availability monitor (Python binding)."""

import logging
import socket
import time
from typing import List, Callable, Tuple

logger = logging.getLogger("nova.mobile.hybrid.device_discovery")

PROBE_TIMEOUT = 1.5
DEFAULT_HOST = "127.0.0.1"
DEFAULT_PORT = 8765


class DeviceDiscovery:
    def __init__(self, host: str = DEFAULT_HOST, port: int = DEFAULT_PORT):
        self.host = host
        self.port = port
        self.is_nova_core_online: bool = False
        self.last_latency_ms: int = -1
        self.consecutive_failures: int = 0
        self._listeners: List[Callable[[bool], None]] = []

    def check_now(self) -> bool:
        reachable, latency = self._probe()
        self.is_nova_core_online = reachable
        self.last_latency_ms = latency
        return reachable

    def simulate_online(self, online: bool = True, latency_ms: int = 12):
        """Used in tests and simulation mode to set Core availability."""
        prev = self.is_nova_core_online
        self.is_nova_core_online = online
        self.last_latency_ms = latency_ms if online else -1
        if prev != online:
            for l in self._listeners:
                l(online)

    def add_state_listener(self, listener: Callable[[bool], None]):
        self._listeners.append(listener)

    def _probe(self) -> Tuple[bool, int]:
        try:
            start = time.time()
            with socket.create_connection((self.host, self.port), timeout=PROBE_TIMEOUT):
                elapsed_ms = int((time.time() - start) * 1000)
                self.consecutive_failures = 0
                return True, elapsed_ms
        except Exception:
            self.consecutive_failures += 1
            return False, -1

File: mobile/hybrid/DeviceDiscovery.kt
package com.nova.mobile.hybrid

import android.util.Log
import kotlinx.coroutines.*
import java.net.InetSocketAddress
import java.net.Socket

/**
 * DeviceDiscovery — Continuously monitors Nova Core (laptop) availability.
 * Checks TCP reachability, tracks latency, and fires callbacks on state changes.
 * Supports auto-reconnect.
 */
class DeviceDiscovery {
    companion object {
        private const val TAG = "DeviceDiscovery"
        private const val PROBE_INTERVAL_MS = 5000L
        private const val PROBE_TIMEOUT_MS = 1500
        private const val DEFAULT_HOST = "192.168.1.100"
        private const val DEFAULT_PORT = 8765
    }

    var host: String = DEFAULT_HOST
    var port: Int = DEFAULT_PORT

    var isNovaCoreOnline: Boolean = false
        private set
    var lastLatencyMs: Long = -1L
        private set
    var consecutiveFailures: Int = 0
        private set

    private val listeners = mutableListOf<(Boolean) -> Unit>()
    private var probeJob: Job? = null

    fun startProbing(scope: CoroutineScope = CoroutineScope(Dispatchers.IO)) {
        probeJob = scope.launch {
            while (isActive) {
                val (reachable, latency) = probe()
                val changed = reachable != isNovaCoreOnline
                isNovaCoreOnline = reachable
                lastLatencyMs = latency
                if (reachable) consecutiveFailures = 0 else consecutiveFailures++
                if (changed) {
                    Log.i(TAG, "Nova Core state changed: ${if (reachable) "ONLINE" else "OFFLINE"} (latency=${latency}ms)")
                    notifyListeners(reachable)
                }
                delay(PROBE_INTERVAL_MS)
            }
        }
    }

    fun stopProbing() {
        probeJob?.cancel()
        probeJob = null
    }

    fun checkNow(): Boolean {
        val (reachable, latency) = probe()
        isNovaCoreOnline = reachable
        lastLatencyMs = latency
        return reachable
    }

    fun addStateListener(listener: (Boolean) -> Unit) {
        listeners.add(listener)
    }

    private fun probe(): Pair<Boolean, Long> {
        return try {
            val start = System.currentTimeMillis()
            Socket().use { socket ->
                socket.connect(InetSocketAddress(host, port), PROBE_TIMEOUT_MS)
            }
            val elapsed = System.currentTimeMillis() - start
            Pair(true, elapsed)
        } catch (e: Exception) {
            Pair(false, -1L)
        }
    }

    private fun notifyListeners(isOnline: Boolean) {
        listeners.forEach { it(isOnline) }
    }
}
